diff --git a/include/u.h b/include/u.h
index 489b2a3886..b5a90b48ef 100644
--- a/include/u.h
+++ b/include/u.h
@@ -85,7 +85,7 @@ typedef jmp_buf sigjmp_buf;
 #endif
 typedef long p9jmp_buf[sizeof(sigjmp_buf)/sizeof(long)];
 
-#if defined(__linux__)
+#if defined(__linux__) || defined(__serenity__)
 #	include <sys/types.h>
 #	if defined(__Linux26__)
 #		include <pthread.h>
diff --git a/src/cmd/dist/a.h b/src/cmd/dist/a.h
index 288063b94b..1906345eee 100644
--- a/src/cmd/dist/a.h
+++ b/src/cmd/dist/a.h
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
+#undef bool
 typedef int bool;
 
 // The Time unit is unspecified; we just need to
diff --git a/src/cmd/dist/unix.c b/src/cmd/dist/unix.c
index 4a78684b44..8c9b24031e 100644
--- a/src/cmd/dist/unix.c
+++ b/src/cmd/dist/unix.c
@@ -10,7 +10,6 @@
 #ifndef WIN32
 #ifndef PLAN9
 
-#include "a.h"
 #include <unistd.h>
 #include <dirent.h>
 #include <sys/stat.h>
@@ -25,6 +24,7 @@
 #include <stdarg.h>
 #include <setjmp.h>
 #include <signal.h>
+#include "a.h"
 
 // bprintf replaces the buffer with the result of the printf formatting
 // and returns a pointer to the NUL-terminated buffer contents.
@@ -673,7 +673,7 @@ main(int argc, char **argv)
 	run(&b, nil, 0, "sysctl", "machdep.cpu.extfeatures", nil);
 	if(contains(bstr(&b), "EM64T"))
 		gohostarch = "amd64";
-#elif defined(__linux__)
+#elif defined(__linux__) || defined(__serenity__)
 	gohostos = "linux";
 #elif defined(__DragonFly__)
 	gohostos = "dragonfly";
@@ -775,6 +775,7 @@ xsamefile(char *f1, char *f2)
 	return streq(f1, f2); // suffice for now
 }
 
+typedef jmp_buf sigjmp_buf;
 sigjmp_buf sigill_jmpbuf;
 static void sigillhand(int);
 
diff --git a/src/lib9/dirfwstat.c b/src/lib9/dirfwstat.c
index 4666e21079..5b45d0e768 100644
--- a/src/lib9/dirfwstat.c
+++ b/src/lib9/dirfwstat.c
@@ -31,7 +31,7 @@ THE SOFTWARE.
 #include <sys/time.h>
 #include <sys/stat.h>
 
-#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__linux__)
+#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__linux__) || defined(__serenity__)
 /* do nothing -- futimes exists and is fine */
 
 #elif defined(__SunOS5_9__)
